src/neuroglancer/datasource/computed/frontend.ts

```diff
@@ -0,0 +1,258 @@
+/**
+ * @license
+ * Copyright 2018 Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ChunkManager, WithParameters} from 'neuroglancer/chunk_manager/frontend';
+import {DataSource, DataSourceProvider, GetVolumeOptions} from 'neuroglancer/datasource';
+import {ComputationParameters, ComputedVolumeChunkSourceParameters} from 'neuroglancer/datasource/computed/base';
+import {DataType, VolumeChunkSpecification, VolumeSourceOptions, VolumeType} from 'neuroglancer/sliceview/volume/base';
+import {MultiscaleVolumeChunkSource as GenericMultiscaleVolumeChunkSource, VolumeChunkSource} from 'neuroglancer/sliceview/volume/frontend';
+import {CancellationToken} from 'neuroglancer/util/cancellation';
+import {mat4, vec3} from 'neuroglancer/util/geom';
+import {SharedObject} from 'neuroglancer/worker_rpc';
+
+
+class ComputedVolumeChunkSource extends
+(WithParameters(VolumeChunkSource, ComputedVolumeChunkSourceParameters)) {}
+
+export abstract class VolumeComputationFrontend<T extends ComputationParameters> extends
+    SharedObject {
+  /**
+   * Modifes the ComputedVolumeDataSourceParameters as needed and returns
+   * parameters for the backend computation.
+   * @param config Config data supplied from the URL
+   * @param volumes A volumes array as returned by
+   *   GenericMultiscaleVolumeChunkSource.getSources()
+   * @param params volume data source parameters, populated with defaults
+   *   from the native resolution origin source. These are to be modified.
+   * @returns a Promise containing parameters passed to the backend
+   *   computation.
+   */
+  abstract initialize(
+      config: string, volumes: VolumeChunkSource[][],
+      params: ComputedVolumeDataSourceParameters): Promise<T>;
+}
+
+export interface VolumeComputationFrontendProvider<T extends ComputationParameters> {
+  makeComputation(): VolumeComputationFrontend<T>;
```

Comment:
I think it might make sense to have this function do the work of initialize, returning a Promise<VolumeComputationFrontend> (you could give VolumeComputationFrontend a constructor that takes the parameters) --- that would be a bit more consistent with how DataSourceProvider works.

Reply:
Done.